parquet/src/main/java/org/apache/iceberg/parquet/ParquetWriter.java

```diff
@@ -185,9 +190,21 @@ public List<Long> splitOffsets() {
     return null;
   }
 
+  /*
+   * Data size could reduce after written out due to encoding/compression.
+   * Use the ratio totalRowGroupSize / totalBufferSize to estimate the size after write out.
+   */
+  private long estimateBufferedSize() {
+    if (totalRowGroupSize == 0 || totalRawBufferedSize == 0 || currentRawBufferedSize == 0) {
+      return writeStore.getBufferedSize();
+    }
+
+    return currentRawBufferedSize * totalRowGroupSize / totalRawBufferedSize;
+  }
+
   private void checkSize() {
     if (recordCount >= nextCheckRecordCount) {
-      long bufferedSize = writeStore.getBufferedSize();
+      long bufferedSize = estimateBufferedSize();
       double avgRecordSize = ((double) bufferedSize) / recordCount;
 
       if (bufferedSize > (targetRowGroupSize - 2 * avgRecordSize)) {
```

Comment:
Do we still need the -2 * avg records size ? Seems like that shouldn't really matter now that we are more accurately estimating our buffered size?